From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ReleaseRaptor <ReleaseRaptor@cixtech.com>
Date: Wed, 19 Feb 2025 14:11:48 +0800
Subject: [PATCH] net: realtek: r8126: Fix a invalid access

Uplayer access pcie memory space of r8126 when pci is not available,
which leads to a system hang. To avoid it, block the access when
carrier is off.

Call trace:
 panic+0x1b8/0x3d8
 nmi_panic+0xac/0xb0
 arm64_serror_panic+0x90/0xa0
 arm64_is_fatal_ras_serror+0x84/0x8c
 do_serror+0x34/0x60
 el1h_64_error_handler+0x30/0x4c
 el1h_64_error+0x64/0x68
 rtl8125_dump_tally_counter+0x78/0xb0 [r8125]
 rtl8125_get_stats64+0x58/0x90 [r8125]
 dev_get_stats+0x4c/0x150
 rtnl_fill_stats+0x44/0x140
 rtnl_fill_ifinfo+0x658/0x1160
 rtmsg_ifinfo_build_skb+0xac/0x120
 rtmsg_ifinfo+0x4c/0xa0
 netdev_state_change+0x7c/0x84
 linkwatch_do_dev+0xac/0xfc
 __linkwatch_run_queue+0xe8/0x230
 linkwatch_event+0x30/0x40
 process_one_work+0x1e0/0x350
 worker_thread+0x6c/0x430
 kthread+0x108/0x10c
 ret_from_fork+0x10/0x20

Signed-off-by: ZHANG Yuntian <yt@radxa.com>
---
 src/r8126_n.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/r8126_n.c b/src/r8126_n.c
index 0c6567f65793..f45e10cda8d3 100644
--- a/src/r8126_n.c
+++ b/src/r8126_n.c
@@ -644,6 +644,10 @@ int rtl8126_dump_tally_counter(struct rtl8126_private *tp, dma_addr_t paddr)
         u32 cmd;
         u32 WaitCnt;
         int retval = -1;
+        struct net_device *dev = tp->dev;
+
+        if (!netif_carrier_ok(dev))
+                return retval;
 
         RTL_W32(tp, CounterAddrHigh, (u64)paddr >> 32);
         cmd = (u64)paddr & DMA_BIT_MASK(32);
-- 
2.47.1

